app/models/manageiq/providers/hawkular/middleware_manager/refresh_parser.rb

```diff
@@ -40,7 +42,54 @@ def get_middleware_servers
             @data[:middleware_servers] << server
             @data_index.store_path(:middleware_servers, :by_ems_ref, server[:ems_ref], server)
           end
-        end.flatten
+        end
+      end
+
+      def fetch_domains_with_servers
+        @data[:middleware_domains] = []
+        @data[:middleware_server_groups] = []
+        @ems.feeds.each do |feed|
+          @ems.domains(feed).each do |domain|
+            parsed_domain = parse_middleware_domain(domain)
+            fetch_server_groups(feed, domain)
+
+            # add the server groups to the domain
+            parsed_domain[:middleware_server_groups] = @data[:middleware_server_groups]
+            @data[:middleware_domains] << parsed_domain
+            @data_index.store_path(:middleware_domains, :by_ems_ref, parsed_domain[:ems_ref], parsed_domain)
+
+            # now it's safe to fetch the domain servers
+            fetch_domain_servers(domain)
+          end
+        end
+      end
+
+      def fetch_server_groups(feed, domain)
+        @ems.server_groups(feed, domain).each do |group|
+          parsed_group = parse_middleware_server_group(group)
+          @data[:middleware_server_groups] << parsed_group
+          @data_index.store_path(:middleware_server_groups, :by_name, parsed_group[:name], parsed_group)
+        end
+      end
+
+      def fetch_domain_servers(domain)
+        @ems.child_resources(domain.path).each do |child|
+          next unless child.type_path.end_with?(hawk_escape_id('Domain WildFly Server'))
+          @eaps << child
+
+          server_name = parse_domain_server_name(child.id)
+          server = parse_middleware_server(child, server_name)
+
+          # add the association to server group (it's well hidden in the inventory)
+          config_path = child.path.to_s.sub(/%2Fserver%3D/, '%2Fserver-config%3D')
+          config = @ems.inventory_client.get_config_data_for_resource(config_path)
+          server_group_name = config['value']['Server Group']
```

Comment:
> is it possible to have 2 server groups with same name within same domain

no, it would be broken configuration of the domain, so the servers would not even start
 
> (or across domains)?

yes, this is possible. Because the server group is not a "top lvl" entity, but it's part of the domain, it shouldn't be a problem. It's identity is determined by the domain and the name